### DISCUSSION TITLE
ArrayBuilder (C++) - How to go back / move the index

### DISCUSSION THREAD

**Discussion by lobis** [scikit-hep/awkward#2903] [Q&A]
I'm using ArrayBuilder on some C++ code to build awkward arrays.

I'm struggling to go back on an append / begin_list call. From some offline conversation with @jpivarski I thought it should be possible to "move the pointer / index" back and keep building (without any overhead, because you are not really deleting anything) but I'm not finding how to do this.

```cpp
template<class PRIMITIVE>
using NumpyBuilder = awkward::LayoutBuilder::Numpy<PRIMITIVE>;

template<class PRIMITIVE, class BUILDER>
using ListOffsetBuilder = awkward::LayoutBuilder::ListOffset<PRIMITIVE, BUILDER>;

ListOffsetBuilder<int, NumpyBuilder<int>> builder;

builder.begin_list();
builder.append(1);
builder.append(2);
builder.being_list();

// I no longer want this list and wish to keep on building
```

Thanks!

**Comment by agoose77** [comment]
@lobis I never spent a huge amount of time thinking about `LayoutBuilder`, but AFAICR we don't implement an "undo". It wouldn't be *that* hard to do (I don't _think_). Is there a reason you can't avoid the list altogether?

**Reply by jpivarski** [reply]
Oh, yeah—this came out of a conversation that @lobis and I had a few days ago. LayoutBuilders don't have an undo, but easily could. (Same for ArrayBuilder, but this is for a LayoutBuilder application that wants to produce arrays without large entries, without knowing ahead of time whether the next entry will be large.)
